### ISSUE TITLE
Is this project still active?

### ISSUE THREAD

**Issue by Taconut** [analog-nico/quota#1]
I'm developing a program in node and I need something exactly like this. If you haven't abandoned this, I'd rather wait for you to get it working than fail at trying to make my own. However, if you aren't planning on finishing this, I'll try to find an alternative (which doesn't seem to be easy).
Sorry to bother you.

**Comment by analog-nico** [comment]
Actually, quota is something I desperately need. My priorities only pushed it further and further down the road. What is your schedule? How long can you wait?

**Comment by Taconut** [comment]
There's no rush for the project. You can continue to procrastinate on it, since my project is currently only privately used by a couple dozen people or so. Its only a pet project, so I don't have a deadline. But quota would be necessary to implement before I release the project publicly.

**Comment by analog-nico** [comment]
Awesome, I will let you know in this issue once I release the first version on NPM.

**Comment by analog-nico** [comment]
Just for my reference. These are @Taconut 's requirements: http://developer.echonest.com/docs/v4#rate-limits

**Comment by analog-nico** [comment]
Hi @Taconut version 0.0.1 is ready and ships with an Echonest preset!

Do I assume correctly that you run your software with a single node.js instance? (I designed Quota to cover cluster environments as well. However, this part is not ready yet but the transition will be easy.)

You may use the following code when you boot your software:
``` js
var quota = require('quota');

var quotaServer = new quota.Server();
quotaServer.addManager('echonest'); // Loads the preset

var quotaClient = new quota.Client(quotaServer);
```

And this code when you want to make a request to Echonest:
``` js
quotaClient.requestQuota('echonest', {}, { requests: 1 }, { maxWait: 10000 })
    .then(function (grant) {

        // Make your request to Echonest here

        // And afterwards do:
        grant.dismiss({ forRule: { main: {
            limit: <X-RateLimit-Limit that was returned in the response header>
        } } })
    })
    .catch(quota.OutOfQuotaError, function (err) {
        // The quota request was denied. E.g. notify the user to try again later.
    })
    .catch(function (err) {
        // Due to technical reasons the quota request was neither granted nor denied. E.g. notify the admins.
    });
```

Quota has internal queueing. So if right now the quota is used up it will wait up to 10 seconds (`maxWait` parameter) for quota to become available.

Since Echonest provides a dynamic rate limit per minute you can feed back the X-RateLimit-Limit header through `grant.dismiss(...)`.

If by any chance your requests fail with a 429 then Quota would activate a backoff mechanism. This is not yet implemented, though. But will be shortly and will require you to pass the 429 info to `grant.dismiss(...)` as well.

You might think Quota's API looks a bit complicated. So true, but Quota is not a slipper. It is a hiking boot. Just compare the presets of [Echonest](https://github.com/analog-nico/quota/blob/master/lib/server/core/presets/echonest.js) and [Google Analytics](https://github.com/analog-nico/quota/blob/master/lib/server/core/presets/google-analytics.js). ;D

If you would integrate it into your software and give me some feedback that would be awesome! The code is well covered by automated tests and if anything comes up I'll give you premium support to find a solution.

**Comment by analog-nico** [comment]
Please use version 0.0.2. It fixes a potentially fatal race condition.

**Comment by Taconut** [comment]
![](http://i.lvme.me/1jj2hk1.jpg)

Wow! You delivered! This API is built beautifully. I have at least three projects that could directly benefit from quota.

Thank you so much for this. May I recommend that you try to spread this project across Reddit, blogs, etc? I feel like you could get quite a bit of reputation for this kind of thing. Its really the only library of its kind currently available.

**Comment by analog-nico** [comment]
:D Thanks! I will definitely spread the word about this library once it reaches a more developed state.

For which APIs would you like to use Quota in your projects? I could provide some presets for you.

**Comment by Taconut** [comment]
Well I was able to inject Quota into the part of my program and it seems to be working fine! Despite what you said, it only took a few minutes to get everything running smoothly. 

As for the other presets, you actually don't have to implement them, since one would be for Google Analytics and the other would be for use within my application.

I do have one question about the soon-to-be backoff mechanism: Will you automatically be re-calling the request after the backoff has cooled down? Because I'd like to know if its going to be necessary to add some sort of logic into the outOfQuotaError function call.

**Comment by analog-nico** [comment]
You will have to tell Quota to activate the backoff mechanism and recall your request yourself. Most likely like this:

``` js
function awesomeEchonestCall() {

   quotaClient.requestQuota('echonest', {}, { requests: 1 }, { maxWait: 10000 })
        .then(function (grant) {

            // You make your request here which gets a 429 from Echonest.

            grant.dismiss({
                backoff: true // <-- THIS TELLS QUOTA TO BACKOFF
            });

            // Since your request failed do your Echonest call again.
            awesomeEchonestCall();
        });

}
```

While the backoff mechanism is active all `.requestQuota(...)` calls will be queued until either `maxWait` elapses and an OutOfQuotaError is thrown or the backoff mechanism cools down and quota grants are returned.

You won't have to add any code to the `.catch(quota.OutOfQuotaError, function (err) { ... })` part. Since queueing quota requests is an integral part you should rather increase `maxWait`.

I hope this is clear enough. Any thought?

**Comment by Taconut** [comment]
I've already implemented the lines you described. Like I said before, everything is extremely intuitive and clear.

One feature that may be of use would be to define a custom BackOff either based on UTC time or a custom inputted value. For example, many APIs such as Vimeo, Github, and Twitter include a "X-RateLimit-Reset" parameter containing the amount of seconds to back off. Additionally, TheEchonest resets every UTC minute, so I could potentially pass "60 - new Date().getUTCSeconds()" as the Backoff.

**Comment by analog-nico** [comment]
Yes, the backoff mechanism will be highly customizable. Often it is just a timeout but some require exponential backoff. (YouTube) Often the code has to backoff altogether and sometimes just for certain API endpoints. (Twitter) The presets - which are the layer on top - will take care of everything so that only certain values like X-RateLimit-Reset need to be passed to `grant.dismiss(...)`.

Regarding Echonest's reset every UTC minute, where can I find the documentation for this? If we know that they really reset at second zero and the servers have a well synchronized time then the backoff only need to wait until the beginning of the next minute like you say. Otherwise the backoff needs to wait exactly 60 seconds whenever a 429 is returned.

**Comment by Taconut** [comment]
In inquired about that in an email to echonest. I'll let you know the details once they respond.

**Comment by analog-nico** [comment]
Perfect. Thank you!
